File: public/sdk/SsoSekolahClient.php
<?php

declare(strict_types=1);

namespace SsoSekolah;

/**
 * ============================================================
 * SSO Sekolah - Client SDK untuk PHP
 * Versi    : 1.0.0
 * Lisensi  : MIT / Open Source
 * ============================================================
 */
class SsoSekolahClient
{
    private string $baseUrl;
    private string $apiKey;
    private int $timeout;

    /**
     * Inisialisasi SDK SSO Sekolah.
     *
     * @param string $baseUrl URL basis API SSO (contoh: https://sso.sekolah.sch.id/api)
     * @param string $apiKey Kunci API Rahasia Aplikasi Anda
     * @param int $timeout Batas waktu request HTTP dalam detik (default: 10)
     */
    public function __construct(string $baseUrl, string $apiKey, int $timeout = 10)
    {
        $this->baseUrl = rtrim($baseUrl, '/');
        $this->apiKey  = $apiKey;
        $this->timeout = $timeout;
    }

    /**
     * Uji koneksi dan validitas API Key ke portal SSO.
     */
    public function testConnection(): array
    {
        return $this->request('GET', '/v1/test');
    }

    /**
     * Ambil daftar pengguna/member dengan filter opsional (role, kelas_id, q, per_page, page).
     */
    public function getMembers(array $params = []): array
    {
        return $this->request('GET', '/v1/members', $params);
    }

    /**
     * Ambil detail data spesifik pengguna berdasarkan UUID/ID.
     */
    public function getMemberById(string $id): array
    {
        return $this->request('GET', "/v1/members/{$id}");
    }

    /**
     * Ambil daftar kelas sekolah (opsional filter: aktif=true, tahun_pelajaran_id).
     */
    public function getKelas(array $params = []): array
    {
        return $this->request('GET', '/v1/kelas', $params);
    }

    /**
     * Ambil daftar tahun pelajaran sekolah (opsional filter: is_aktif=true).
     */
    public function getTahunPelajaran(array $params = []): array
    {
        return $this->request('GET', '/v1/tahun-pelajaran', $params);
    }

    /**
     * Ambil daftar peran (roles) yang terdaftar di SSO.
     */
    public function getPeran(): array
    {
        return $this->request('GET', '/v1/data/peran');
    }

    /**
     * Ambil data statistik agregat jumlah pengguna & aplikasi.
     */
    public function getStatistik(): array
    {
        return $this->request('GET', '/v1/data/statistik');
    }

    /**
     * Dekode & Verifikasi Payload Token JWT SSO tanpa pustaka luar.
     */
    public static function verifyJwtToken(string $jwtToken): ?array
    {
        $parts = explode('.', $jwtToken);
        if (count($parts) !== 3) {
            return null;
        }

        $payload = json_decode(base64_decode(strtr($parts[1], '-_', '+/')), true);
        if (!is_array($payload)) {
            return null;
        }

        // Periksa expired token jika ada klaim 'exp'
        if (isset($payload['exp']) && time() >= $payload['exp']) {
            return null; // Token kedaluwarsa
        }

        return $payload;
    }

    /**
     * Eksekusi HTTP Client menggunakan cURL native.
     */
    private function request(string $method, string $path, array $params = []): array
    {
        $url = $this->baseUrl . $path;
        if (!empty($params)) {
            $url .= '?' . http_build_query($params);
        }

        $ch = curl_init();
        curl_setopt_array($ch, [
            CURLOPT_URL            => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT        => $this->timeout,
            CURLOPT_HTTPHEADER     => [
                'X-API-Key: ' . $this->apiKey,
                'Accept: application/json',
            ],
            CURLOPT_SSL_VERIFYPEER => false,
        ]);

        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        $error    = curl_error($ch);
        curl_close($ch);

        if ($error) {
            throw new \RuntimeException("Gagal menghubungkan ke SSO: " . $error);
        }

        $data = json_decode((string) $response, true);
        if ($httpCode >= 400) {
            $msg = $data['pesan'] ?? "Error API SSO (HTTP Status {$httpCode})";
            throw new \RuntimeException($msg);
        }

        return $data ?? [];
    }
}

File: public/sdk/sso_sekolah_client.py
"""
============================================================
SSO Sekolah - Client SDK untuk Python 3
Versi    : 1.0.0
Lisensi  : MIT / Open Source
============================================================
"""

import json
import urllib.parse
import urllib.request
import base64
import time

class SsoSekolahClient:
    def __init__(self, base_url: str, api_key: str, timeout: int = 10):
        """
        Inisialisasi Client SDK SSO Sekolah.
        :param base_url: URL basis API SSO (contoh: https://sso.sekolah.sch.id/api)
        :param api_key: Kunci API rahasia aplikasi Anda
        :param timeout: Batas waktu request dalam detik
        """
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key
        self.timeout = timeout

    def test_connection(self) -> dict:
        """Uji koneksi & validitas API Key."""
        return self._request('GET', '/v1/test')

    def get_members(self, **params) -> dict:
        """Ambil daftar pengguna/member SSO dengan filter opsional."""
        return self._request('GET', '/v1/members', params)

    def get_member_by_id(self, member_id: str) -> dict:
        """Ambil detail data spesifik pengguna berdasarkan ID/UUID."""
        return self._request('GET', f'/v1/members/{member_id}')

    def get_kelas(self, **params) -> dict:
        """Ambil daftar kelas sekolah."""
        return self._request('GET', '/v1/kelas', params)

    def get_tahun_pelajaran(self, **params) -> dict:
        """Ambil daftar tahun pelajaran sekolah."""
        return self._request('GET', '/v1/tahun-pelajaran', params)

    def get_peran(self) -> dict:
        """Ambil daftar peran/roles yang terdaftar."""
        return self._request('GET', '/v1/data/peran')

    def get_statistik(self) -> dict:
        """Ambil data statistik agregat pengguna & aplikasi."""
        return self._request('GET', '/v1/data/statistik')

    @staticmethod
    def verify_jwt_token(jwt_token: str) -> dict:
        """Dekode & Verifikasi Payload Token JWT SSO tanpa library external."""
        try:
            parts = jwt_token.split('.')
            if len(parts) != 3:
                return None
            padded = parts[1] + '=' * (-len(parts[1]) % 4)
            payload_bytes = base64.urlsafe_b64decode(padded)
            payload = json.loads(payload_bytes.decode('utf-8'))

            if 'exp' in payload and time.time() >= payload['exp']:
                return None  # Expired
            return payload
        except Exception:
            return None

    def _request(self, method: str, path: str, params: dict = None) -> dict:
        url = f"{self.base_url}{path}"
        if params:
            filtered = {k: v for k, v in params.items() if v is not None}
            if filtered:
                url += '?' + urllib.parse.urlencode(filtered)

        headers = {
            'X-API-Key': self.api_key,
            'Accept': 'application/json',
            'User-Agent': 'SsoSekolahPythonSDK/1.0'
        }

        req = urllib.request.Request(url, headers=headers, method=method)
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as response:
                body = response.read().decode('utf-8')
                return json.loads(body)
        except urllib.error.HTTPError as e:
            error_body = e.read().decode('utf-8')
            try:
                err_json = json.loads(error_body)
                raise Exception(err_json.get('pesan', f"HTTP Error {e.code}"))
            except json.JSONDecodeError:
                raise Exception(f"HTTP Error {e.code}: {error_body}")
        except Exception as e:
            raise Exception(f"Gagal menghubungkan ke SSO: {str(e)}")
